File: graphs/hasCycle.js
/**
 * @param {number} numCourses
 * @param {number[][]} prerequisites
 * @return {boolean}
 */
 var canFinish = function(numCourses, prerequisites) {
    var graph = new Array(numCourses);
    for (var i = 0; i < numCourses; i++) graph[i] = [];
    var len = prerequisites.length;
    for (var i = 0; i < len; i++) {
        var item = prerequisites[i];
        graph[item[0]].push(item[1]);
    }
    
    // empty: other
    // 1: visiting
    // 2: visited
    let statuses = new Array(numCourses);
    const hasCycle = (course) => {
        if(statuses[course] == 1) return true;
        if(statuses[course] == 2) return false;
        statuses[course] = 1;
        for (let c of graph[course]) {
            if(hasCycle(c)) return true;
        }
        statuses[course] = 2;
    }
    
    for(let i = 0; i<numCourses; i++) {
        if(hasCycle(i)) {
            return false;
        }
    }
    return true;
};

File: graphs/hasCycle.go
package graphs


func canFinish(numCourses int, prerequisites [][]int) bool {
    graph := make([][]int, numCourses)
    for i:=0; i<len(prerequisites); i++ {
        item:= prerequisites[i]
        graph[item[0]] = append(graph[item[0]], item[1])
    }

    statuses := make([]int, numCourses)
    
    for i := 0; i<numCourses; i++ {
        if hasCycle(i, &statuses, &graph) {
            return false
        }
    }
    
    return true
}

func hasCycle(course int, statuses *[]int, graph *[][]int) bool {
    if (*statuses)[course] == 1 {
        return true
    }
    if (*statuses)[course] == 2 {
        return false
    }
    (*statuses)[course] = 1
    for _, v := range (*graph)[course] {
        if hasCycle(v, statuses, graph) {
            return true
        }
    }
    (*statuses)[course] = 2
    return false
}